### PR TITLE
Support PathLike inputs.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

My take on PathLike support.
Alternate for #8481; marked as release critical only in the sense that one of the two PRs should go in...

Also xref https://github.com/matplotlib/matplotlib/pull/10048.

## PR Checklist

- [X] Has Pytest style unit tests
- [X] Code is PEP 8 compliant
- [X] New features are documented, with examples if plot related
- [X] Documentation is sphinx and numpydoc compliant
- [X] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by anntzer** [comment]
changed file -> fh.

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -1349,11 +1349,12 @@ def imsave(fname, arr, vmin=None, vmax=None, cmap=None, format=None,
...
-    # Fast path for saving to PNG
-    if (format == 'png' or format is None or
-            isinstance(fname, six.string_types) and
-            fname.lower().endswith('.png')):
+    if isinstance(fname, getattr(os, "PathLike", ())):
+        fname = os.fspath(fname)
+    if (format == 'png'
+        or (format is None
+            and isinstance(fname, six.string_types)
+            and fname.lower().endswith('.png'))):
```
Comment (QuLogic) [inline lib/matplotlib/image.py]: This precedence is something changed in another PR, right?

Comment (QuLogic) [reply lib/matplotlib/image.py]: Ah, #10048, I see you linked it.

File: doc/users/next_whats_new/2018-01-10-AL-pathlike.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+PathLike support
+````````````````
+
+The following functions now support `~.PathLike` inputs on Python 3.6+:
```
[the remark this answers is not in the document]
Comment (anntzer) [reply doc/users/next_whats_new/2018-01-10-AL-pathlike.rst]: fixed

File: doc/users/next_whats_new/2018-01-10-AL-pathlike.rst [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+On Python 3.6+, `~matplotlib.pyplot.savefig`, `~matplotlib.pyplot.imsave`,
+`~matplotlib.pyplot.imread`, and animation writers now accept `~os.PathLike`\s
```
[the remark this answers is not in the document]
Comment (anntzer) [reply doc/users/next_whats_new/2018-01-10-AL-pathlike.rst]: fixed

**Comment by anntzer** [comment]
fixed in 2.2.1 by https://github.com/matplotlib/matplotlib/pull/10754.

**Pull request merged**